### PR TITLE
document public API

### PR DESCRIPTION
**Pull request by artemredkin**

### CODE REVIEW

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 1]
```diff
@@ -67,6 +103,12 @@ public class HTTPClient {
         }
     }
 
+    /// Execute POST request using specified URL.
+    ///
+    /// - parameters:
+    ///     - url: Remote URL.
+    ///     - body: Request body.
+    ///     - deadline: Maximum request duration.
```
Comment (weissi) [inline Sources/AsyncHTTPClient/HTTPClient.swift]: I wouldn't document a dealine as a duration. I'd maybe write the time when the request must have been completed by or so.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 2]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
-/// released together with the HTTPTaskHandler when channel is closed
+/// HTTPClientResponseDelegate allows to receive notification about request processing and to control how reponse parts are processed.
+///
+///  - note: This delegate is strongly held by the `HTTPTaskHandler`
+///          for the duration of the `Request` processing and will be
+///          released together with the `HTTPTaskHandler` when channel is closed.
 public protocol HTTPClientResponseDelegate: AnyObject {
     associatedtype Response
 
+    /// Called when request `Head` is sent. Will be called once.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: Called when the request head is sent. Will be called once.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 3]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
+    ///
+    /// - parameters:
+    ///     - task: Current request context.
+    ///     - head: Request `Head`.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: No backticks.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 4]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
     }
 }
 
-/// This delegate is strongly held by the HTTPTaskHandler
-/// for the duration of the HTTPRequest processing and will be
-/// released together with the HTTPTaskHandler when channel is closed
+/// HTTPClientResponseDelegate allows to receive notification about request processing and to control how reponse parts are processed.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: Backticks around type names.

I think this needs a more user-centric explanation, something like: "You can implement this protocol if you need fine control over an HTTP request/response, for example if you want to inspect the response headers before deciding whether to accept a response body, or if you want to stream your request body. Pass an instance of your conforming class to the `HTTPClient.execute()` method and this package will call each delegate method appropriately as the request takes place."

Just a thought.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 5]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
     associatedtype Response
 
+    /// Called when request `Head` is sent. Will be called once.
+    ///
+    /// - parameters:
+    ///     - task: Current request context.
+    ///     - head: Request `Head`.
     func didSendRequestHead(task: HTTPClient.Task<Response>, _ head: HTTPRequestHead)
 
+    /// Called when request part was sent. Could be called zero or more times.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: "Called when a part of the request body is sent. Could be called zero or more times."

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 6]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
 
+    /// Called when response `Head` is received. Will be called once. Response processing will resume when returned `EventLoopFuture` is fullfilled.
+    ///
+    /// - parameters:
+    ///     - task: Current request context.
+    ///     - head: Received reposonse head.
+    /// - returns: `EventLoopFuture` that will be used for backpressure.
     func didReceiveHead(task: HTTPClient.Task<Response>, _ head: HTTPResponseHead) -> EventLoopFuture<Void>
 
+    /// Called when response body `Part` is received. Could be called zero or more times. Response processing will resume when returned `EventLoopFuture` is fullfilled.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: "Called when part of a response body is received. Could be called zero or more times. You must return an `EventLoopFuture<Void>` that you complete when you have finished processing the body part. You can create an already succeeded future by calling `task.eventLoop.makeSucceededFuture(())`."

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 7]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
+    ///
+    /// - parameters:
+    ///     - task: Current request context.
+    ///     - buffer: Received body `Part`.
+    /// - returns: `EventLoopFuture` that will be used for backpressure.
     func didReceivePart(task: HTTPClient.Task<Response>, _ buffer: ByteBuffer) -> EventLoopFuture<Void>
 
+    /// Called when error was thrown during request execution. Request processing will be stopped.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: Will be called zero or one time only?

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 8]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
+    ///
+    /// - parameters:
+    ///     - task: Current request context.
+    ///     - error: Error that occured during response processing.
     func didReceiveError(task: HTTPClient.Task<Response>, _ error: Error)
 
+    /// Called when response is fully processed. Client is requred to return result of the processing specified by `Response` associated type.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: "Called when the complete HTTP request is finished. You must return an instance of your `Response` associated type. Will be called once, except if an error occurred.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 9]
```diff
@@ -238,7 +328,9 @@ internal extension URL {
 }
 
 extension HTTPClient {
+    /// Response execution context.
```
Comment (ianpartridge) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: Worth explaining that the user does not instantiate this themselves; the package does it.

File: Sources/AsyncHTTPClient/HTTPClient.swift (hunk 1 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 10]
```diff
@@ -16,16 +16,30 @@ import Foundation
 import NIOHTTP1
 
 extension HTTPClient {
+    /// A representation of an HTTP cookie.
     public struct Cookie {
+        /// The name of the cookie.
         public var name: String
+        /// The cookie's sting value.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 11]
```diff
@@ -16,16 +16,30 @@ import Foundation
...
         public var value: String
+        /// The cookie's path.
         public var path: String
+        /// The domain of the cookie
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 12]
```diff
@@ -16,16 +16,30 @@ import Foundation
...
         public var domain: String?
+        /// The cookie's expiration date.
         public var expires: Date?
+        /// The cookie's age.
         public var maxAge: Int?
+        /// A Boolean value that indicates whether the cookie should only be sent to HTTP servers.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 13]
```diff
@@ -16,16 +16,30 @@ import Foundation
...
         public var httpOnly: Bool
+        /// A Boolean value that indicates whether this cookie should only be sent over secure channels.
         public var secure: Bool
 
+        /// Parses HTTP cookie from `Set-Cookie` header.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 5 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 6 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 8 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 9 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: Is this better?

**Comment by artemredkin** [comment]
@ianpartridge thank you!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 9 above)
Comment (ianpartridge) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: It looks the same to me?

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 14]
```diff
@@ -191,24 +243,62 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
     }
 }
 
-/// This delegate is strongly held by the HTTPTaskHandler
-/// for the duration of the HTTPRequest processing and will be
-/// released together with the HTTPTaskHandler when channel is closed
+/// HTTPClientResponseDelegate allows to receive notification about request processing and to control how reponse parts are processed.
+///
+///  - note: This delegate is strongly held by the `HTTPTaskHandler`
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: I think this is important for implementors of the protocol.

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 15]
```diff
@@ -16,16 +16,30 @@ import Foundation
...
         public var path: String
+        /// The domain of the cookie
         public var domain: String?
+        /// The cookie's expiration date.
         public var expires: Date?
+        /// The cookie's age.
         public var maxAge: Int?
+        /// A Boolean value that indicates whether the cookie should only be sent to HTTP servers.
         public var httpOnly: Bool
+        /// A Boolean value that indicates whether this cookie should only be sent over secure channels.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 16]
```diff
@@ -59,13 +84,21 @@ extension HTTPClient {
...
+        /// Remote HTTP scheme, resolved from `URL`.
         public var scheme: String
+        /// Remote host, resolved from `URL`.
         public var host: String
+        /// Request custom HTTP Headers, defaults to no headers.
         public var headers: HTTPHeaders
+        /// Request body, defaults to no body.
         public var body: Body?
 
         public init(url: String, version: HTTPVersion = HTTPVersion(major: 1, minor: 1), method: HTTPMethod = .GET, headers: HTTPHeaders = HTTPHeaders(), body: Body? = nil) throws {
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 7 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 9 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: Fixed, sorry, forgot to push.

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 17]
```diff
@@ -16,16 +16,30 @@ import Foundation
...
         public var secure: Bool
 
+        /// Parses HTTP cookie from `Set-Cookie` header.
+        ///
+        /// - parameters:
+        ///     - string: String representation of the `Set-Cookie` response header.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: done, thanks!

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 18]
```diff
@@ -18,12 +18,42 @@ import NIOConcurrencyHelpers
...
+/// HTTPClient class provides API for request execution.
+///
+/// Example:
+///
+/// ```swift
+///     let client = HTTPClient(eventLoopGroupProvider = .createNew)
+///     client.get(url: "https://swift.org", deadline: .now() + .seconds(1)).whenComplete { result in
+///         switch result {
+///         case .failure(let error):
+///         // process error
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 19]
```diff
@@ -18,12 +18,42 @@ import NIOConcurrencyHelpers
...
+/// ```swift
+///     try client.syncShutdown()
+/// ```
 public class HTTPClient {
     public let eventLoopGroup: EventLoopGroup
     let eventLoopGroupProvider: EventLoopGroupProvider
     let configuration: Configuration
     let isShutdown = Atomic<Bool>(value: false)
 
+    /// Create an HTTPClient with specified `EventLoopGroup` provider and configuration.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: done!

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 20]
```diff
@@ -58,6 +89,11 @@ public class HTTPClient {
         }
     }
 
+    /// Execute GET request using specified URL.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 21]
```diff
@@ -187,10 +257,15 @@ public class HTTPClient {
         }
     }
 
+    /// HTTPClient configuration.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 22]
```diff
@@ -187,10 +257,15 @@ public class HTTPClient {
...
     public struct Configuration {
+        /// TLS configuration, defaults to `TLSConfiguration.forClient()`.
         public var tlsConfiguration: TLSConfiguration?
+        /// Enables following certain 3xx redirects automatically, defaults to `false`.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: added description, thanks!

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 23]
```diff
@@ -208,15 +283,26 @@ public class HTTPClient {
...
     public enum EventLoopGroupProvider {
+        /// `EventLoopGroup` will be provided by the user. Owner of this group is responsible for it's lifecycle.
         case shared(EventLoopGroup)
+        /// `EventLoopGroup` will be created by the client. When `syncShutdown` is called, created `EventLoopGroup` will be shut down as well.
         case createNew
     }
 
+    /// Timeout configuration
     public struct Timeout {
+        /// Specifies `connect` timeout.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: done!

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 24]
```diff
@@ -20,28 +20,49 @@ import NIOSSL
 
 extension HTTPClient {
 
+    /// Represent request body.
     public struct Body {
+        /// Chunk provider.
         public struct StreamWriter {
             let closure: (IOData) -> EventLoopFuture<Void>
 
+            /// Write data to upstream connection.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 25]
```diff
@@ -20,28 +20,49 @@ import NIOSSL
...
+            ///
+            /// - parameters:
+            ///     - data: Data to write.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 26]
```diff
@@ -20,28 +20,49 @@ import NIOSSL
...
             public func write(_ data: IOData) -> EventLoopFuture<Void> {
                 return self.closure(data)
             }
         }
 
+        /// Body size.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: Added description about optionality. Not sure about `var` - user can provide incorrect pair of stream and length anyway... There is no way to validate it anyway...

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 27]
```diff
@@ -76,6 +109,18 @@ extension HTTPClient {
             try self.init(url: url, version: version, method: method, headers: headers, body: body)
         }
 
+        /// Create HTTP request.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 3 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (ianpartridge) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: Can you explain why? Maybe there's a way to phrase this information that makes it more actionable by the user, for example: "This package will hold a strong reference to your delegate instance for the duration of the execution of your request. You are not required to maintain a reference yourself, but may do so if you wish".

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 2 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 4 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: Great idea, added!

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 28]
```diff
@@ -253,14 +368,20 @@ extension HTTPClient {
             self.lock = Lock()
         }
 
+        /// `EventLoopFuture` of this request.
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]: Should this be something like: "`EventLoopFuture` for the response returned by this request"?

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 29]
```diff
@@ -90,6 +105,17 @@ extension HTTPClient {
...
 
+        /// Create HTTP cookie.
+        ///
+        /// - parameters:
+        ///     - name: The name of the cookie.
+        ///     - value: The cookie's string value.
+        ///     - path: The cookie's path.
+        ///     - domain: The domain of the cookie, defaults to nil.
+        ///     - expires: The cookie's expiration date, defaults to nil.
+        ///     - maxAge: The cookie's age, defaults to nil.
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: maximum age in ... (seconds?)

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 30]
```diff
@@ -90,6 +104,17 @@ extension HTTPClient {
             }
         }
 
+        /// Create HTTP cookie.
+        ///
+        /// - parameters:
+        ///     - name: The name of the cookie.
+        ///     - value: The cookie's sting value.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: Good idea, added some clarification, tanks!

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift [hunk 31]
```diff
@@ -16,18 +16,33 @@ import Foundation
...
         public var name: String
+        /// The cookie's string value.
         public var value: String
+        /// The cookie's path.
         public var path: String
+        /// The domain of the cookie.
         public var domain: String?
+        /// The cookie's expiration date.
         public var expires: Date?
+        /// The cookie's age.
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 28 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift (hunk 29 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift]: fixed, thanks!

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 32]
```diff
@@ -208,15 +292,26 @@ public class HTTPClient {
         }
     }
 
+    /// Specifies how `EventLoopGroup` will be created and establishes lifecycle ownership.
     public enum EventLoopGroupProvider {
+        /// `EventLoopGroup` will be provided by the user. Owner of this group is responsible for it's lifecycle.
```
Comment (Yasumoto) [inline Sources/AsyncHTTPClient/HTTPClient.swift]:
```suggestion
        /// `EventLoopGroup` will be provided by the user. Owner of this group is responsible for its lifecycle.
```

Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: Thank you!

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 33]
```diff
@@ -191,24 +258,71 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
     }
 }
 
-/// This delegate is strongly held by the HTTPTaskHandler
-/// for the duration of the HTTPRequest processing and will be
-/// released together with the HTTPTaskHandler when channel is closed
+/// `HTTPClientResponseDelegate` allows to receive notification about request processing and to control how reponse parts are processed.
```
Comment (Yasumoto) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]:
```suggestion
/// `HTTPClientResponseDelegate` allows an implementation to receive notifications about request processing and to control how response parts are processed.
```

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 34]
```diff
@@ -191,24 +258,71 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
+/// You can implement this protocol if you need fine control over an HTTP request/response, for example if you want to inspect the response
```
Comment (Yasumoto) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]:
```suggestion
/// You can implement this protocol if you need fine-grained control over an HTTP request/response, for example, if you want to inspect the response
```

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 35]
```diff
@@ -191,24 +258,71 @@ internal class ResponseAccumulator: HTTPClientResponseDelegate {
...
+/// headers before deciding whether to accept a response body, or if you want to stream your request body. Pass an instance of your conforming
+/// class to the `HTTPClient.execute()` method and this package will call each delegate method appropriately as the request takes place.
+///
+///  - note: This delegate is strongly held by the `HTTPTaskHandler`
```
Comment (Yasumoto) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]:
```suggestion
///  - note: This delegate is strongly held by the `HTTPTaskHandler`.
```

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 36]
```diff
@@ -238,7 +352,10 @@ internal extension URL {
 }
 
 extension HTTPClient {
+    /// Response execution context. Will be created by the library and could be used for obtaining
+    /// `EventLoopFuture<Response>` of the execution or cancellation of the execution.
     public final class Task<Response> {
+        /// `EventLoop` used to execute and process request.
```
Comment (Yasumoto) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]:
```suggestion
        /// `EventLoop` used to execute and process requests.
```

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 37]
```diff
@@ -253,14 +370,20 @@ extension HTTPClient {
             self.lock = Lock()
         }
 
+        /// `EventLoopFuture` for the response returned by this request.
         public var futureResult: EventLoopFuture<Response> {
             return self.promise.futureResult
         }
 
+        /// Waits for execution of request.
```
Comment (Yasumoto) [inline Sources/AsyncHTTPClient/HTTPHandler.swift]:
```suggestion
        /// Waits for execution of this request to complete.
```

**Comment by Yasumoto** [comment]
Feel free to add/ignore those small tweaks; this looks amazing and is super helpful! 🥇

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 35 above)
Comment (ianpartridge) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: This is the middle of a sentence so a full stop doesn't make sense here?

**Comment by ianpartridge** [comment]
Thanks @Yasumoto! I committed the obvious fixes :)

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 35 above)
Comment (Yasumoto) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: Whoa, yep. Somehow totally missed the entire paragraph below! 😅

Thanks!

Comment (ianpartridge) [reply Sources/AsyncHTTPClient/HTTPHandler.swift]: No problem - thanks for reviewing all this!

**Pull request merged**